transports/websocket/src/framed.rs

```diff
@@ -442,55 +442,68 @@ fn location_to_multiaddr<T>(location: &str) -> Result<Multiaddr, Error<T>> {
 
 /// The websocket connection.
 pub struct Connection<T> {
-    receiver: BoxStream<'static, Result<IncomingData, connection::Error>>,
+    receiver: BoxStream<'static, Result<Incoming, connection::Error>>,
     sender: Pin<Box<dyn Sink<OutgoingData, Error = connection::Error> + Send>>,
     _marker: std::marker::PhantomData<T>
 }
 
-/// Data received over the websocket connection.
+/// Data or control information received over the websocket connection.
 #[derive(Debug, Clone)]
-pub enum IncomingData {
-    /// Binary application data.
-    Binary(Vec<u8>),
-    /// UTF-8 encoded application data.
-    Text(Vec<u8>),
+pub enum Incoming {
+    /// Application data.
+    Data(Data),
     /// PONG control frame data.
-    Pong(Vec<u8>)
+    Pong(Vec<u8>),
+    /// Close reason.
+    Closed(CloseReason),
```

Comment:
To me, it is still a bit confusing why this is treated as an `Event` and thus returned via `Incoming` instead of being treated as an error and thus be returned via `Error::Closed`.

I am not familiar enough with soketto, thus I don't think this should block the effort.

Background for other readers: https://github.com/paritytech/soketto/pull/29#pullrequestreview-633290924.

Reply:
It's not clear cut what the semantics should be. I went for the `Error::Closed` approach at first too, but @tomaka [pointed out](https://github.com/paritytech/soketto/pull/29#pullrequestreview-633290924) that the closing of the connection does carry information (a code in the simple case, but the specs allow for essentially any amount of "extra data").
The use case I have in mind for this is "collaborative backoff" where friendly peers can use the close reason to ask for respite, e.g. substrate telemetry asking nodes to send less data or go away for a while. In that context I think making the close reason an `Error` might be unfortunate.